src/mpi/coll/reduce/reduce.c

```diff
@@ -321,6 +334,94 @@ int MPIR_Reduce_impl(const void *sendbuf, void *recvbuf, int count,
     goto fn_exit;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIR_Reduce_decision_tree
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_Reduce_decision_tree(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
+                              MPI_Op op, int root, MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t size;
+    cJSON *device_info;
+
+    /* TODO - Calculate size of buffer */
+
+    mpi_errno = MPIU_Decision_tree_traverse("reduce", comm_ptr, size, datatype, coll_info);
+
+    if (unlikely(mpi_errno != MPI_SUCCESS))
+        MPIR_ERR_POP(mpi_errno);
+    else {
+        cJSON *mpir = cJSON_GetObjectItemCaseSensitive(coll_info, "mpir");
+
+        /* If the JSON tree is using an MPIR-level collective, parse that information here. */
+        if (likely(cJSON_IsString(mpir) && 0 != strcmp(mpir->valuestring, "device"))) {
```

Comment:
What is `mpir`?

Reply:
☝️ `cJSON *mpir`

It's just the `mpir` part of the JSON object.